internal/backend/local/test.go

```diff
@@ -309,43 +272,127 @@ func (runner *TestFileRunner) Test(file *moduletest.File) {
+			// For now, this means a hard stop has been requested, in this case we don't
+			// even stop to mark future test runs as having been skipped. They'll
 			// just show up as pending in the printed summary. We will quickly
 			// just mark the overall file status has having errored to indicate
 			// it was interrupted.
-			file.Status = file.Status.Merge(moduletest.Error)
 			return
 		}
 
+		// the walkFn is called asynchronously, and needs to be recovered
+		// separately in the case of a panic.
+		defer logging.PanicHandler()
+
+		log.Printf("[TRACE] vertex %q: starting visit (%T)", dag.VertexName(v), v)
+
+		defer func() {
+			if r := recover(); r != nil {
+				// If the walkFn panics, we get confusing logs about how the
+				// visit was complete. To stop this, we'll catch the panic log
+				// that the vertex panicked without finishing and re-panic.
+				log.Printf("[ERROR] vertex %q panicked", dag.VertexName(v))
+				panic(r) // re-panic
+			}
+
+			if diags.HasErrors() {
+				for _, diag := range diags {
+					if diag.Severity() == tfdiags.Error {
+						desc := diag.Description()
+						log.Printf("[ERROR] vertex %q error: %s", dag.VertexName(v), desc.Summary)
+					}
+				}
+				log.Printf("[TRACE] vertex %q: visit complete, with errors", dag.VertexName(v))
+			} else {
+				log.Printf("[TRACE] vertex %q: visit complete", dag.VertexName(v))
+			}
+		}()
+
+		// Acquire a lock on the semaphore
+		sem.Acquire()
+		defer sem.Release()
+
+		switch v := v.(type) {
+		case *graph.NodeTestRun:
+			// TODO: The execution of a NodeTestRun is currently split between
+			// its Execute method and the continuation of the walk callback.
+			// Eventually, we should move all the logic related to a test run into
+			// its Execute method, effectively ensuring that the Execute method is
+			// enough to execute a test run in the graph.
+			diags = v.Execute(runner.EvalContext)
+			if diags.HasErrors() {
+				return diags
+			}
+			// continue
+		case graph.GraphNodeExecutable:
+			diags = v.Execute(runner.EvalContext)
+			return diags
+		default:
+			// If the vertex isn't a test run or executable, we'll just skip it.
+			return
+		}
+
+		// We already know that the vertex is a test run
+		runNode := v.(*graph.NodeTestRun)
+
+		file := runNode.File()
+		run := runNode.Run()
+
 		if runner.Suite.Stopped {
 			// Then the test was requested to be stopped, so we just mark each
 			// following test as skipped, print the status, and move on.
 			run.Status = moduletest.Skip
 			runner.Suite.View.Run(run, file, moduletest.Complete, 0)
-			continue
+			return
 		}
-
+		file.Lock()
 		if file.Status == moduletest.Error {
 			// If the overall test file has errored, we don't keep trying to
 			// execute tests. Instead, we mark all remaining run blocks as
 			// skipped, print the status, and move on.
 			run.Status = moduletest.Skip
 			runner.Suite.View.Run(run, file, moduletest.Complete, 0)
-			continue
+			file.Unlock()
+			return
 		}
+		file.Unlock()
```

Comment:
should we move this above the `.Execute()` calls above? Just so we check if we should execute before we do any of the work?

Reply:
Sure. It was only okay to do this because most of the execution is still done in this file, but that may change in future, so will move 👍 
=> 28e93e4a29